pkg/networkservice/chains/client/client.go

```diff
@@ -48,32 +52,74 @@ type Option func(c *clientOptions)
 		c.name = name
-	})
+	}
 }
 
 // WithAdditionalFunctionality sets additionalFunctionality for the client. Note: this adds into tail of the client chain.
 func WithAdditionalFunctionality(additionalFunctionality ...networkservice.NetworkServiceClient) Option {
-	return Option(func(c *clientOptions) {
+	return func(c *clientOptions) {
 		c.additionalFunctionality = additionalFunctionality
-	})
+	}
 }
 
 // WithAuthorizeClient sets authorizeClient for the client chain.
 func WithAuthorizeClient(authorizeClient networkservice.NetworkServiceClient) Option {
 	if authorizeClient == nil {
 		panic("authorizeClient cannot be nil")
 	}
-	return Option(func(c *clientOptions) {
+	return func(c *clientOptions) {
 		c.authorizeClient = authorizeClient
-	})
+	}
 }
 
 // NewClient - returns a (1.) case NSM client.
-//             - ctx    - context for the lifecycle of the *Client* itself.  Cancel when discarding the client.
-//             - cc - grpc.ClientConnInterface for the endpoint to which this client should connect
-func NewClient(ctx context.Context, cc grpc.ClientConnInterface, clientOpts ...Option) networkservice.NetworkServiceClient {
+//  - ctx - context for the lifecycle of the *Client* itself.  Cancel when discarding the client.
+func NewClient(ctx context.Context, connectTo *url.URL, dialOptions []grpc.DialOption, clientOpts ...Option) networkservice.NetworkServiceClient {
+	var rv networkservice.NetworkServiceClient
+	var opts = &clientOptions{
+		name:            "client-" + uuid.New().String(),
+		authorizeClient: null.NewClient(),
+	}
+	for _, opt := range clientOpts {
+		opt(opts)
+	}
+	rv = chain.NewNetworkServiceClient(
+		append(
+			append([]networkservice.NetworkServiceClient{
+				updatepath.NewClient(opts.name),
+				serialize.NewClient(),
+				refresh.NewClient(ctx),
+				metadata.NewClient(),
+			}, opts.additionalFunctionality...),
+			opts.authorizeClient,
+			adapters.NewServerToClient(
+				chain.NewNetworkServiceServer(
+					clienturl.NewServer(connectTo),
+					heal.NewServer(ctx, nil),
+					connect.NewServer(
+						ctx,
+						func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
+							return chain.NewNetworkServiceClient(
+								heal.NewClient(ctx, networkservice.NewMonitorConnectionClient(cc)),
+								networkservice.NewNetworkServiceClient(cc),
+							)
+						},
+						connect.WithDialOptions(dialOptions...),
+					),
+				),
+			),
+		)...,
+	)
+	return rv
+}
+
+// NewServerSideClient returns an NSM client for (2.) and (3.) cases.
+// This function is intended to be a part of a client factory, and therefore should never ba called directly.
+//  ctx - context for the lifecycle of the *Client* itself.  Cancel when discarding the client.
+//  cc  - grpc.ClientConnInterface for the endpoint to which this client should connect.
+func NewServerSideClient(ctx context.Context, cc grpc.ClientConnInterface, clientOpts ...Option) networkservice.NetworkServiceClient {
```

Comment:
I am deeply uncertain that introducing a function like this is a good idea to solve our problem

Reply:
Could you elaborate on your concerns, please?

Reply:
I agreed with @edwarnicke 

@d-uzlov Can we just inline this into NewClientFactory